text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,391 @@
+It feels like integration tests should only validate functionality we shouldn't be breaking (cdk command line options, docker build command, and other user facing behavior).
+
+There are definitely cases where we "break" some behavior, that isn't necessarily impacting users that much, which is why its ok to do.
+
+For example, the [simplify docker asset publishing](https://github.com/aws/aws-cdk/pull/5733) PR removed the ECRAdoptedRepository construct, which is technically a breaking change. The failing tests due to that, are unit tests, not integration tests (need to double check). My argument is that if we did have an integration test that validates the existence of an `ECRAdoptedRepository`, that validation probably didn't belong their in the first place. In essence, it means the integration test was validating the implementation rather than the expected result, which is that images are properly treated as assets (uploaded to somewhere you can download it from).
+
+#### Caveat
+
+Even if we do agree that such a scenario means the test should be changed, how would we actually escape the failing test? We cannot change the previous test...
+
+## Avoiding Regressions
+
+You might ask yourself, assuring compatibility should surely involve regression testing. That is, running **old** integration tests on a **new** cli version and a **new** framework version. You are corret, however, this issue is out of scope for this RFC. 
+
+**We are only trying to solve CLI vs Framework Compatibility issues**
+
+A discussion about regressions can be found [here](https://github.com/aws/aws-cdk/issues/5962), and will probably evolve into an RFC of its own.
+
+## Assuring Multi Version Compatibility
+
+The proposed design only validates compatibility to the **last** previous version. However, does this ensure compatibility across multiple versions? Specifically, lets talk about CLI compatibility to older framework versions.
+
+Imagine version `1.19.0` performs some kind of schema evolution with respect to version `1.18.0`.
+
+```typescript
+if (manifest.version === '1.18.0') {
+
+    manifest = justUpgradeVersion(manifest, '1.19.0');
+
+    manifest.superImportant = 'some-critical-default'
+
+}
+```
+
+Now we want to release version `1.20.0`, our compatibility test makes sure that version `1.20.0` is compatible to version `1.19.0`. Does that mean it is compatible with version `1.18.0` as well?
+
+Well, yes! 
+
+If version `1.19.0` is detected, we are compatible, and if version `1.18.0` is detected, we perform an evolution to `1.19.0` and everything should work the same.
+
+However, what if a bug is introduced, that now makes the code look like:
+
+```typescript
+if (manifest.version === '1.18.0') {
+    manifest = justUpgradeVersion(manifest, '1.19.0');
+}
+```
+
+In this case, our compatibility test still passes, because version `1.19.0` has the `superImportant` property, but compatibility to version `1.18.0` is broken.
```

Comment:
Even if we only go back one version, wouldn't we have detected this problem when running the backwards compat test when we released `1.19.0`? Wouldn't that have entailed reading an assembly with version `1.18.0` which was missing the essential property?

Reply:
No, because the bug, which is removing this line: 

```typescript
manifest.superImportant = 'some-critical-default'
```

might have only been introduced in version `1.20.0`.

Reply:
Hm. Well that shouldn't(tm) happen, but I see what you mean.